### PR TITLE
feat: `tbl_format_setup()` gains a `setup` argument that supports printing the header before the data for the body is available, e.g., for remote backends such as databases

### PR DESCRIPTION
**Pull request by krlmlr**
This supports printing the header before the body or footer is ready. Adding the `setup` argument to `tbl_format_setup()` was the easiest way I found to achieve this. The two commits are self-contained.

### CODE REVIEW

**Comment by krlmlr** [comment]
This is likely to affect dbplyr by no longer showing the number of rows in the header if it's less than 21. I'd trade that for showing the header first. And I wonder if we should temporarily show the structure of the data too, and erase it (with `\r`?) when the body is ready.

File: R/tbl-format.R [hunk 1]
```diff
@@ -60,17 +60,39 @@ format_tbl <- function(
...
-  setup <- tbl_format_setup(x,
-    width = width, ...,
+  setup_used <- FALSE
+
+  setup <- tbl_format_setup(
+    x,
+    width = width,
+    ...,
+    setup = {
+      setup_used <- TRUE
```
Comment (DavisVaughan) [inline R/tbl-format.R:70]: I'd add a note in here about the fact that "If `setup` is evaluated by the `tbl_format_setup()` method, then that's our signal to recall `tbl_format_setup()` a second time"

File: R/tbl-format.R [hunk 2]
```diff
@@ -60,17 +60,39 @@ format_tbl <- function(
...
     n = n,
     max_extra_cols = max_extra_cols,
     max_footer_lines = max_footer_lines,
     focus = attr(x, "pillar_focus")
   )
 
-  header <- tbl_format_header(x, setup)
-  body <- tbl_format_body(x, setup)
-  footer <- tbl_format_footer(x, setup)
+  header <- transform(tbl_format_header(x, setup))
```
Comment (DavisVaughan) [inline R/tbl-format.R:79]: What does `transform()` do?

File: R/tbl-format-setup.R [hunk 3]
```diff
@@ -41,6 +41,14 @@
 #'   This argument is mandatory for all implementations of this method.
 #' @param ...
 #'   Extra arguments to [print.tbl()] or [format.tbl()].
+#' @param setup
```
Comment (DavisVaughan) [inline R/tbl-format-setup.R:44]: Probably worth a news bullet

**Review by DavisVaughan — COMMENTED** [review]
This feels quite "spooky" to me

In particular I found relying on "did `setup` get evaluated or not" a bit hard to understand and document.

Is it possible that `new_tbl_format_setup()` could gain a boolean argument that identifies whether or not this is a "partial" setup?

Like:
- If we see `is.null(setup)` then we do partial setup and return with `partial = TRUE`.
- That causes `tbl_format_setup()` to get called again, but with `!is.null(setup)` this time (i.e. it gets the results of the original partial setup), so we do the full setup and return with `partial = FALSE`

File: R/tbl-format.R (hunk 2 above)
Comment (krlmlr) [reply R/tbl-format.R:79]: `transform` is `writeLines` when printing and `identity` when formatting. This is how we achieve incremental printing.

Adding a comment to that effect to the code.

**Comment by krlmlr** [comment]
Thanks. I see how it's spooky, I couldn't think of a better way.

We want to be compatible with existing _implementations_ of `tbl_format_setup()` who might not know or care about the new flow. We effectively implement a "pre-setup" (for just the header) and a "setup for real" (for the body and footer). Should we do a new generic instead, or perhaps two new generics? What would they look like, what would the migration path look like?

**Comment by krlmlr** [comment]
Going with this implementation for now, we can revisit as needed.

**Pull request merged**